### PR TITLE
[CPU] [Feat]  Add native AMX-FP8 attention impl for Diamond Rapids

### PR DESCRIPTION
**Pull request by zhejiangxiaomai**
```markdown
## Purpose

Add a native AMX-FP8 attention path for Diamond Rapids CPUs.

The existing AMX path (added in 22524f7a92) dequants K/V to BF16 before MMA even when FP8 KV-cache is used. This PR introduces a dedicated `AMX_FP8` ISA that uses hardware FP8 MMA intrinsics (`_tile_dpfp8ps` for QK, `_tile_dpbf16ps` for PV) to avoid the dequant overhead. `_get_attn_isa` automatically selects `amx_fp8` over `amx` when FP8 KV-cache is active and the hardware supports it.

Also required: bump Dockerfile base to Ubuntu 25.04 + GCC-15 (binutils ≥ 2.44 is needed to assemble AMX-FP8 mnemonics), and add `-fno-tree-pre` for GCC ≥ 15 to work around a GCC ICE in oneDNN.

## Test Plan

```bash
docker build --target vllm-test -f docker/Dockerfile.cpu \
  -t vllm-cpu-amx-fp8:gcc15-ubuntu2504 .

docker run --rm vllm-cpu-amx-fp8:gcc15-ubuntu2504 \
  pytest -q tests/kernels/attention/test_cpu_attn.py -k "amx_fp8 or amx" -x
```

## Test Result

```
594 passed, 2070 deselected in 455s
```

rename_deepklox

## Performance Results

| Workload | Before | After |
|---|---:|---:|
| Decode, KV128 | 0.75x | **1.14x** |
| Decode, KV8K | 1.58x | **1.61x** |
| Prefill, Q128/KV128 | 0.30x | **0.77x** |
| Prefill, Q512/KV512 | 0.56x | **0.96x** |
| Prefill, Q128/KV8K | 1.12x | **1.15x** |

Speedups are relative to BF16 AMX. The optimization vectorizes dynamic Q scaling and uses hardware-assisted E4M3 quantization.

**Correctness:** 82 passed, 27 skipped.

### CODE REVIEW

**Comment by bigPYJ1151** [comment]
Hi @zhejiangxiaomai Can we do some kernel benchmark with `benchmarks/kernels/cpu/benchmark_cpu_attn.py` to check the improvement of pure FP8 AMX compared with BF16 AMX.

And we should also do some E2E acc test like https://github.com/vllm-project/vllm/pull/39445 did.

File: csrc/cpu/cpu_attn_impl.hpp [hunk 1]
```diff
@@ -1019,8 +1024,16 @@ class AttentionMainLoop {
           attention_impl_t::v_cache_token_group_stride(block_size);
       // v_cache_head_group_stride: stride of V cache when move to next
       // HeadDimAlignment head dims in a block
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_impl.hpp:1026]: This comment should be moved to the above of `v_cache_head_group_stride `

File: csrc/cpu/generate_cpu_attn_dispatch.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 # ISA types
 ISA_TYPES = {
     "AMX": 0,
+    "AMX_FP8": 7,
```
Comment (bigPYJ1151) [inline csrc/cpu/generate_cpu_attn_dispatch.py]: Move it to the tail

File: docker/Dockerfile.cpu [hunk 3]
```diff
@@ -140,7 +146,7 @@ RUN apt-get update -y \
 WORKDIR /workspace
 
 COPY requirements/build/rust.txt requirements/build/rust.txt
-RUN python3 -m pip install --no-cache-dir -r requirements/build/rust.txt
+RUN python3 -m pip install --break-system-packages --no-cache-dir -r requirements/build/rust.txt
```
Comment (bigPYJ1151) [inline docker/Dockerfile.cpu:168]: Is this necessary?

File: csrc/cpu/cpu_attn_amx_fp8.hpp [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+        for (int64_t e = 0; e < fp8_per_row; ++e) {
+          row_base[e] = quant_fn(static_cast<float>(ksrc[s * fp8_per_row + e]),
+                                 k_inv);
+        }
+      }
+    }
+  }
+}
+
+#if __GNUC__ >= 15
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_amx_fp8.hpp]: Let's remove such compiler version checks and remove the fallback path. We can always assume the toolchain supports the latest features. And these new features are only supported on DMR.

File: csrc/cpu/cpu_attn_amx_fp8.hpp [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+//   E5M2 V cache: falls back to BF16×FP8 dequant path (TODO: native E5M2).
+// ---------------------------------------------------------------------------
+template <typename kv_cache_t>
+class TileGemm224<uint8_t, kv_cache_t> {
+  static_assert(std::is_same_v<kv_cache_t, c10::Float8_e4m3fn> ||
+                    std::is_same_v<kv_cache_t, c10::Float8_e5m2>,
+                "kv_cache_t must be Float8_e4m3fn or Float8_e5m2");
+
+  // For PV E5M2 fallback: BF16 probs × FP8 E5M2 V (dequant to BF16).
+  using pv_gemm_fallback_t = TileGemm224<c10::BFloat16, kv_cache_t>;
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_amx_fp8.hpp]: We should also add AMX path for E5M2 instead of fallback to BF16

File: csrc/cpu/cpu_attn_amx_fp8.hpp [hunk 6]
```diff
@@ -0,0 +1,887 @@
...
+           (head_dim % 64 != 0))
+class AttentionImpl<ISA::AMX_FP8, scalar_t, head_dim, kv_cache_scalar_t>
+    : public AttentionImpl<ISA::AMX, scalar_t, head_dim, kv_cache_scalar_t> {
+};
+
+template <typename scalar_t, int64_t head_dim, typename kv_cache_scalar_t>
+  requires(std::is_same_v<scalar_t, c10::BFloat16> &&
+           std::is_same_v<kv_cache_scalar_t, c10::Float8_e4m3fn> &&
+           (head_dim % 64 == 0))
+class AttentionImpl<ISA::AMX_FP8, scalar_t, head_dim, kv_cache_scalar_t> {
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_amx_fp8.hpp:588]: Is it necessary to inherit from the AMX impl, I think almost all methods need to override.

File: csrc/cpu/cpu_attn_amx_fp8.hpp [hunk 7]
```diff
@@ -0,0 +1,887 @@
...
+  // Q buffer holds FP8-quantized query.
+  using query_t = scalar_t;         // BF16 input
+  using q_buffer_t = uint8_t;       // FP8 quantized, stored in Q buffer
+  using kv_cache_t = kv_cache_scalar_t;
+  using logits_buffer_t = float;
+  using partial_output_buffer_t = float;
+  using prob_buffer_t = scalar_t;   // BF16 softmax probs (PV phase)
+
+  constexpr static int64_t BlockSizeAlignment =
+      std::is_same_v<kv_cache_t, c10::Float8_e4m3fn> ? 64 : 32;
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_amx_fp8.hpp]: It can be hardcode as `64`.

File: docker/Dockerfile.cpu (hunk 3 above)
Comment (zhejiangxiaomai) [reply docker/Dockerfile.cpu:168]: --break-system-packages is required because Ubuntu 25.04 enforces PEP 668 for its system Python. Removing it would cause the pip install command in the Rust build stage to fail.

File: csrc/cpu/cpu_attn_impl.hpp (hunk 1 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_impl.hpp:1026]: fixed.

File: csrc/cpu/cpu_attn_amx_fp8.hpp (hunk 4 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_amx_fp8.hpp]: fixed.

File: csrc/cpu/cpu_attn_amx_fp8.hpp (hunk 6 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_amx_fp8.hpp:588]: fixed.

File: csrc/cpu/cpu_attn_amx_fp8.hpp (hunk 7 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_amx_fp8.hpp]: fixed.

File: csrc/cpu/cpu_attn_amx_fp8.hpp (hunk 5 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_amx_fp8.hpp]: fixed.

File: csrc/cpu/generate_cpu_attn_dispatch.py (hunk 2 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/generate_cpu_attn_dispatch.py]: fixed.

File: csrc/cpu/cpu_attn_amx_fp8.hpp [hunk 8]
```diff
@@ -0,0 +1,778 @@
...
+  // Q buffer holds FP8-quantized query.
+  using query_t = scalar_t;         // BF16 input
+  using q_buffer_t = uint8_t;       // FP8 quantized, stored in Q buffer
+  using kv_cache_t = kv_cache_scalar_t;
+  using logits_buffer_t = float;
+  using partial_output_buffer_t = float;
+  using prob_buffer_t = scalar_t;   // BF16 softmax probs (PV phase)
+
+  constexpr static int64_t BlockSizeAlignment = 64;
+  constexpr static int64_t HeadDimAlignment = 2 * (AMX_TILE_ROW_BYTES / 4);
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_amx_fp8.hpp]: We can just hardcode this as 64. For AMX-FP8, tile size is `[16, 64]`, so QK phase requires 64 alignment. For PV phase, the output channel number is `2 * (64 / 4) = 32` (2 is because we always use 2 AMX tile on B martix), so  it requires 32 alignment. We can use the common multiple 64 as alignment factor.

File: csrc/cpu/cpu_attn_impl.hpp [hunk 9]
```diff
@@ -1017,10 +1022,18 @@ class AttentionMainLoop {
...
+        }
+        return headdim_alignment;
+      }();
       // v_cache_head_group_stride: stride of V cache when move to next
       // HeadDimAlignment head dims in a block
       const int64_t v_cache_head_group_stride =
-          attention_impl_t::v_cache_head_group_stride(block_size);
+          pv_headdim_alignment == headdim_alignment
+              ? attention_impl_t::v_cache_head_group_stride(block_size)
+              : block_size * pv_headdim_alignment;
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_impl.hpp]: After change the global head size alignment to 64, these changes are not required.

File: csrc/cpu/cpu_attn_amx_fp8.hpp (hunk 8 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_amx_fp8.hpp]: fixed.

File: csrc/cpu/cpu_attn_impl.hpp (hunk 9 above)
Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_impl.hpp]: fixed.

**Comment by bigPYJ1151** [comment]
Thanks @zhejiangxiaomai Overall looks good. Please finally check the accuracy and performance locally because current CI doesn't have DMR instance. And fix some failures in pre-commit check.

**Comment by zhejiangxiaomai** [comment]
> Thanks @zhejiangxiaomai Overall looks good. Please finally check the accuracy and performance locally because current CI doesn't have DMR instance. And fix some failures in pre-commit check.

I checked both accuracy and performance locally on a DMR machine after rebasing onto the latest main.

Accuracy: 55 passed, 54 skipped.

Performance results remain consistent with the pre-rebase results, with no regressions observed. FP8 provides clear gains for long-context decode and prefill, while the known short-prefill overhead remains unchanged.

**Review by bigPYJ1151 — APPROVED** [review]

File: csrc/cpu/cpu_attn_amx.hpp [hunk 10]
```diff
@@ -238,6 +240,8 @@ class TileGemm122<c10::BFloat16, kv_cache_t> {
                     std::is_same_v<kv_cache_t, c10::Float8_e5m2>,
                 "kv_cache_t must be BFloat16, Float8_e4m3fn, or Float8_e5m2");
 
+  static constexpr bool pv_uses_separate_tile_config = false;
+
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_amx.hpp]: Seems we can remove this

Comment (zhejiangxiaomai) [reply csrc/cpu/cpu_attn_amx.hpp]: fixed

**Comment by zhejiangxiaomai** [comment]
Locally validated on Diamond Rapids after rebasing onto the latest `main`:

- Accuracy: `54 passed, 54 skipped`
- Full test suite: `1640 passed, 1044 skipped`
- Performance: ~39% faster for long decode and ~13.5% faster for long-context prefill. Prefill-512 was neutral; short prefill remains slower due to Q quantization overhead.

**Comment by louie-tsai** [comment]
@louie-tsai DMR
